bgpd/bgp_conditional_adv.c

```diff
@@ -0,0 +1,451 @@
+				/* Handle configuration changes */
+				if (peer->advmap_info[afi][safi]
+					    .config_change) {
+					adv_conditional =
+						(filter->advmap.condition
+						 == CONDITION_EXIST)
+							? NLRI
+							: WITHDRAW;
+				} else {
+					if (prev_ret != RMAP_MATCH)
+						adv_conditional =
+							(filter->advmap
+								 .condition
+							 == CONDITION_EXIST)
+								? NLRI
+								: WITHDRAW;
+				}
+				peer->advmap_info[afi][safi].cmap_prev_status =
+					ret;
+				break;
+
+			case RMAP_NOMATCH:
+				/* Handle configuration changes */
+				if (peer->advmap_info[afi][safi]
+					    .config_change) {
+					adv_conditional =
+						(filter->advmap.condition
+						 == CONDITION_EXIST)
+							? WITHDRAW
+							: NLRI;
+				} else {
+					if (prev_ret != RMAP_NOMATCH)
+						adv_conditional =
+							(filter->advmap
+								 .condition
+							 == CONDITION_EXIST)
+								? WITHDRAW
+								: NLRI;
+				}
+				peer->advmap_info[afi][safi].cmap_prev_status =
+					ret;
+				break;
+
+			case RMAP_OKAY:
+			case RMAP_ERROR:
+			default:
+				break;
+			}
+
+			/* amap (route-map attached to advertise-map)
+			 * validation.
+			 */
+			ret = is_rmap_valid(filter->advmap.amap) ? RMAP_MATCH
+								 : RMAP_NOOP;
+
+			if ((ret == RMAP_NOOP) && (prev_ret == RMAP_NOOP))
+				continue;
+
+			/* Derive conditional advertisement status from
+			 * condition and return value of condition-map
+			 * validation.
+			 */
+			if (adv_conditional == NLRI)
+				filter->advmap.status = true;
+			else if (adv_conditional == WITHDRAW)
+				filter->advmap.status = false;
+			else {
+				/* no change in advertise status. So, only
+				 * previously withdrawn routes will be
+				 * advertised if needed.
+				 */
+			}
+
+			/* Send regular update as per the existing policy.
+			 * There is a change in route-map, match-rule, ACLs,
+			 * or route-map filter configuration on the same peer.
+			 */
+			if (peer->advmap_info[afi][safi].config_change) {
+				paf = peer_af_find(peer, afi, safi);
+				if (paf) {
+					update_subgroup_split_peer(paf, NULL);
+					subgrp = paf->subgroup;
+					if (subgrp && subgrp->update_group)
+						subgroup_announce_table(
+							paf->subgroup, NULL);
+				}
+				peer->advmap_info[afi][safi].config_change =
+					false;
+			}
+
+			/* Send update as per the conditional advertisement */
+			if (adv_conditional) {
+				route_advertised = bgp_conditional_adv_routes(
+					peer, afi, safi, table,
+					filter->advmap.amap,
+					filter->advmap.status);
+
+				/* amap_prev_status is only to check whether we
+				 * have announced any routes(advertise/withdraw)
+				 * or not. filter->advmap.status will have the
+				 * actual filter status
+				 */
+				peer->advmap_info[afi][safi].amap_prev_status =
+					route_advertised ? RMAP_MATCH
+							 : RMAP_NOOP;
+			}
+		}
+	}
+	return 0;
+}
+
+void bgp_conditional_adv_enable(struct peer *peer, afi_t afi, safi_t safi)
+{
+	struct bgp *bgp = peer->bgp;
+
+	assert(bgp);
```

Comment:
can this be NULL?

Reply:
probably not, but in case it is it will `assert` (because it is unexpected). I don't see harm in this and it will avoid coverity issue (if any).